lib/topology.ml

```diff
@@ -0,0 +1,250 @@
+let seq_gen_2n n =
+  (* A node can either be present in the output or not,
+   * so use a loop [1, 2^n) and have the [i]th bit determine that.
+   * *)
+  let of_mask i =
+    seq_range 0 n |> Seq.filter (fun bit -> (i lsr bit) land 1 = 1)
+  in
+  seq_range 1 (1 lsl n) |> Seq.map of_mask
+
+(** [seq_sort ~cmp s] sorts [s] in a temporary place using [cmp]. *)
+let seq_sort ~cmp s =
+  let a = Array.of_seq s in
+  Array.fast_sort cmp a ; Array.to_seq a
+
+(** [seq_append a b] is the sequence [a] followed by [b] *)
+let seq_append (a : 'a Seq.t) (b : 'a Seq.t) =
+  let rec next v () =
+    match v () with
+    | Seq.Nil ->
+        b ()
+    | Seq.Cons (x, xs) ->
+        Seq.Cons (x, next xs)
+  in
+  next a
+
+module NUMA = struct
+  type node = Node of int
+
+  (* no mutation is exposed in the interface,
+   * therefore this is immutable *)
+  type t =
+    { distances: int array array
+    ; cpu_to_node: node array
+    ; node_cpus: CPUSet.t array
+    ; all: CPUSet.t
+    ; candidates: (float * node Seq.t) Seq.t }
+
+  let node_of_int i = Node i
+
+  let node_distances d nodes =
+    let dists =
+      nodes
+      |> Seq.flat_map (fun n1 -> nodes |> Seq.map (fun n2 -> d.(n1).(n2)))
+    in
+    let count, max_dist =
+      Seq.fold_left
+        (fun (count, maxv) e -> (count + 1, max maxv e))
+        (0, min_int) dists
+    in
+    let avg_dist = (Seq.fold_left ( + ) 0 dists |> float) /. float count in
+    (* We want to minimize maximum distance first, and average distance next.
+     * When running the VM we don't know which pCPU it'll end up using,
+     * and want to limit the worst case performance.
+     *)
+    ((max_dist, avg_dist), nodes)
+
+  let dist_cmp (a1, _) (b1, _) = compare a1 b1
+
+  let gen_candidates d =
+    (* We fully expand all combinations up to 16 NUMA nodes.
+     * Higher than that we approximate: the node "n" becomes synonymous with
+     * real NUMA nodes [n*multiply ... n*multiply + multiply-1],
+     * except we always the add the single NUMA node combinations.
+     *)
+    let numa_nodes = Array.length d in
+    let gen_n = min 16 numa_nodes in
+    let multiply = numa_nodes / gen_n in
+    if multiply * gen_n <> numa_nodes then
+      invalid_arg
+        (Printf.sprintf "Invalid number of NUMA nodes: %d" numa_nodes) ;
+    (* make sure that single NUMA nodes are always present in the combinations *)
+    let single_nodes =
+      seq_range 0 (Array.length d)
+      |> Seq.map (fun i -> ((10, 10.0), Seq.return i))
+    in
+    gen_n |> seq_gen_2n
+    |> Seq.map (fun nodes ->
+           if multiply = 1 then nodes
+           else
+             nodes
+             |> Seq.flat_map (fun n ->
+                    seq_range 0 multiply
+                    |> Seq.map (fun i -> (n * multiply) + i)))
+    |> Seq.map (node_distances d)
+    |> seq_append single_nodes |> seq_sort ~cmp:dist_cmp
+    |> Seq.map (fun ((_, avg), nodes) ->
+           (avg, Seq.map (fun n -> Node n) nodes))
```

Comment:
this would be more readable if those anonymous functions had names

Reply:
I removed the anonymous functions to simplify this code: when you have >16 NUMA nodes it'll try just the individual NUMA nodes, and if those are all too small fall back to all NUMA nodes.
This simplifies the algorithm. When we have hardware that has >16 NUMA nodes we can reevaluate what the best approach would be.